File: chapter6/steps/steps.cpp
#include <iostream>
#include <vector>

using namespace std;

int main (int argc, char ** argv) {
    int numCases;
    cin >> numCases;
    int currentCase = 1;
    while (currentCase <= numCases) {
        int x, y;
        cin >> x;
        cin >> y;
        int distance = y - x;
        int mid = distance/2;
        // vector<int> steps = new vector<int>(2);
        // steps[0] = 0;
        // steps[1] = 1;
        int stepSize = 1;
        int numSteps = 1;
        distance -= 1;
        while (distance > mid) {
            
            stepSize++;
            distance -= stepSize;
            numSteps++;
            cout << "\tTaken " << numSteps << " steps! Only " << distance << " distance left!\n";
        }
        if (distance == mid) {
            distance -= stepSize;
            numSteps++;
            cout << "\tTaken " << numSteps << " steps! Only " << distance << " distance left!\n";
        }
        while (distance > 0) {
            stepSize--;
            distance -= stepSize;
            numSteps++;
            cout << "\tTaken " << numSteps << " steps! Only " << distance << " distance left!\n";
        }
        cout << numSteps << endl;

        currentCase++;
    }

    return 0;
}

File: chapter06/steps/steps.py
# UVa 846

def main():
    numCases = int(input())
    currentCase = 1
    while currentCase <= numCases:
        start, end = input().split(" ")
        start = int(start)
        end = int(end)
        distance = end - start
        numSteps = 0
        if distance != 0:
            totalDistanceTraveled = 0
            nextStep = 1
            # every other step we take we want to increment the size (simulating stepping from both ends)
            # EX: 1 1 for 2, 1 1 1, for 3 (we are adding on the outter pieces first and working to the mid)
            # while loop can go over (EX: 1 2 1 is still 3 steps for the distance of 3, even though the distance added up is 4)
            count = 0
            while distance > totalDistanceTraveled:
                if count % 2 == 0 and count != 0:
                    nextStep += 1
                totalDistanceTraveled += nextStep
                numSteps += 1
                count += 1
        print(numSteps)
        currentCase += 1

if __name__ == "__main__":
    main()